src/distPlus/symbolic/SymbolicDist.re

```diff
@@ -296,103 +313,407 @@ module GenericSimple = {
+        `RenderedShape(Distributions.Continuous.reduce((+.), [|sc1, sc2|]), Distributions.Discrete.reduce((+.), [|sd1, sd2|]), 1. +. i2)
+    }
+    | (`RenderedShape(sc1, sd1, i1), `RenderedShape(sc2, sd2, i2)) => {
+        `RenderedShape(Distributions.Continuous.reduce((+.), [|sc1, sc2|]), Distributions.Discrete.reduce((+.), [|sd1, sd2|]), i1 +. i2)
+    }
+    }
   };
-};
 
-let toString = (r: bigDist) =>
-  r
-  |> (
-    fun
+  let evaluatePointwiseProduct = (et1: nodeResult, et2: nodeResult, n: int) => {
+    switch ((et1, et2)) {
+    /* Known cases: */
+    | (`Simple(`Float(v1)), `Simple(`Float(v2))) => {
+        v1 == v2
+        ? `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.make({xs: [|v1|], ys: [|1.|]}), 1.)
+        : `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.empty, 0.) // TODO: add warning: shouldn't pointwise multiply scalars.
+    }
+    | (`Simple(`Float(v1)), `Simple(d2)) => {
+        // evaluate d2 at v1
+        let y = GenericSimple.pdf(v1, d2);
+        `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.make({xs: [|v1|], ys: [|y|]}), y)
+    }
+    | (`Simple(d1), `Simple(`Float(v2))) => {
+        // evaluate d1 at v2
+        let y = GenericSimple.pdf(v2, d1);
+        `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.make({xs: [|v2|], ys: [|y|]}), y)
+    }
+    | (`Simple(`Normal(n1)), `Simple(`Normal(n2))) => {
+        let mean = (n1.mean *. n2.stdev**2. +. n2.mean *. n1.stdev**2.) /. (n1.stdev**2. +. n2.stdev**2.);
+        let stdev = 1. /. ((1. /. n1.stdev**2.) +. (1. /. n2.stdev**2.));
+        let integral = 0; // TODO
+        `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.empty, 0.)
+    }
+    /* General cases */
+    | (`Simple(d1), `Simple(d2)) => {
+        // NOT IMPLEMENTED YET
+        // TODO: evaluate integral properly
+        let (sc1, sd1) = renderDistributionToXYShape(d1, n);
+        let (sc2, sd2) = renderDistributionToXYShape(d2, n);
+        `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.empty, 0.)
+    }
+    | (`Simple(d1), `RenderedShape(sc2, sd2, i2)) => {
+        // NOT IMPLEMENTED YET
+        // TODO: evaluate integral properly
+        let (sc1, sd1) = renderDistributionToXYShape(d1, n);
+        `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.empty, 0.)
+    }
+    | (`RenderedShape(sc1, sd1, i1), `Simple(d1)) => {
+        // NOT IMPLEMENTED YET
+        // TODO: evaluate integral properly
+        let (sc2, sd2) = renderDistributionToXYShape(d1, n);
+        `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.empty, 0.)
+    }
+    | (`RenderedShape(sc1, sd1, i1), `RenderedShape(sc2, sd2, i2)) => {
+        `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.empty, 0.)
+    }
+    }
+  };
+
+
+  let evaluateNormalize = (et: nodeResult, n: int) => {
```

Comment:
Do you think there would be realistic cases of people entering n=0? If not, we could have a type of ``nonZeroInt`` , or even ``intOver2`` or similar. Then we could deal with those cases upfront, and not have to need special cases for them internally.

Reply:
Do you mean `n=0` here (i.e. sampleCount) or `i=0`, which is the integral of the subtree to be normalized?

Reply:
``n==0``. You do a check for it here:
```
| `RenderedShape(sc, sd, 0.) => {
        `RenderedShape(Distributions.Continuous.empty, Distributions.Discrete.empty, 0.)
      }
```
That wouldn't be necessary if this were impossible.

Reply:
That zero is for `i`, the third element of a `` `RenderedShape``. It represents the integral of the tree. It's supposed to catch cases like `normalize(scaleBy(normal(0, 1), 0))` or `normalize(pointwiseProduct(uniform(0,1), uniform(2, 3)))`. I don't think that can be dealt with elsewhere, or am I missing something?

Reply:
Ah, I misunderstood what n was for. It was really not intuitive that it meant integral; it's an int, and ``i`` is often used for index or length. 

You may want a custom type for `RenderedShape. Like, `RenderedShape({mixedShape: mixedShape, integralAmount: float})